### PR TITLE
Merging kick_policy into master

### PR DESCRIPTION
**Pull request by alexandremuzio**

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (alexandremuzio) [inline .vscode/settings.json]: @luckeciano can I delete this file here?

Comment (luckeciano) [reply .vscode/settings.json]: Yes, sorry for it

File: baselines/ppo1/mlp_policy.py [hunk 2]
```diff
@@ -40,6 +40,8 @@ def _init(self, ob_space, ac_space, hid_size, num_hid_layers, gaussian_fixed_var
             else:
                 pdparam = tf.layers.dense(last_out, pdtype.param_shape()[0], name='final', kernel_initializer=U.normc_initializer(0.01))
 
+        my_var = tf.strided_slice(mean, [0], [1], [1], shrink_axis_mask=1)
+        my_var_out = tf.identity(my_var, name='output_node')
```
Comment (luckeciano) [inline baselines/ppo1/mlp_policy.py:44]: I think it is important to understand these two lines. Here I basically create an output node from computational graph, which is necessary to freeze the subgraph we want to deploy into the agent.

File: baselines/ppo1/kick_policy.py [hunk 3]
```diff
@@ -12,13 +12,12 @@
 class KickPolicy(object):
```
Comment (luckeciano) [inline baselines/ppo1/kick_policy.py:12]: This code is very messy (sorry for it haha) but at least is very useful when we want to load weights from a pre-trained network using Keras. I spent a good time to integrate everything (I also needed to change other things in the optimization loop), so this is a good starting point if you want to learn a controller by imitation and then optimize it.

File: baselines/ppo1/pposgd_simple.py [hunk 4]
```diff
@@ -3,20 +3,23 @@
...
 import tensorflow as tf, numpy as np
+from tensorflow.python.framework import graph_util, graph_io
 import time
 from baselines.common.mpi_adam import MpiAdam
 from baselines.common.mpi_moments import mpi_moments
 from mpi4py import MPI
 from collections import deque
 
-def traj_segment_generator(pi, env, horizon, stochastic):
+def traj_segment_generator(pi, env, horizon, stochastic, rw_scaler):
```
Comment (luckeciano) [inline baselines/ppo1/pposgd_simple.py:13]: The Reward Scaler is an idea I found in a paper, where the reward is normalized using running statistics (just like the way inputs are normalized). It does not worked much better - actually, just using a multiplication factor was better. But I maintained here because in some problems it can be more helpful.

File: baselines/ppo1/pposgd_simple.py [hunk 5]
```diff
@@ -133,29 +143,34 @@ def learn(env, policy_fn, *,
         for (oldv, newv) in zipsame(oldpi.get_variables(), pi.get_variables())])
     compute_losses = U.function([ob, ac, atarg, ret, lrmult], losses)
 
-    U.initialize()
+    initialize_uninitialized(tf.get_default_session())
```
Comment (luckeciano) [inline baselines/ppo1/pposgd_simple.py:146]: This line is very important to make a pre-trained network works properly. The problem of `U.initialize()` - or even the similar method in tensorflow - is because it initializes the whole graph, making all the weights random again (screwing pre-training stuff). However, if I simply remove it, we do not initialize the exploration parameters, so I created this method that solves this problem.

File: baselines/ppo1/pposgd_simple.py [hunk 6]
```diff
@@ -133,29 +143,34 @@ def learn(env, policy_fn, *,
...
+    #U.initialize()
     adam.sync()
 
     # Load existing model
     if load_model:
         U.load_state(os.path.join(model_dir, "model"))
 
+    # Reinitialize exploration
+    #reinitialize_exploration(tf.get_default_session())
```
Comment (luckeciano) [inline baselines/ppo1/pposgd_simple.py:155]: This is a technique I called exploration restarter. In my experiments, I found that gaussian exploration is the most important hyperparameter, even more than learning rate. However, in PPO it tends to shrink fast and converge prematurely. Thus, I create this exploration restarter which restart exploration rate using the initial value defined in the network, instead of simply load the learned values from previous training. It worked well, mainly when the learning rate and exploration rate do not have the same magnitude. 

As you can see, I let this commented because it is not the default behavior, but it can be improved with a simple flag.

File: baselines/ppo1/run_soccer.py [hunk 7]
```diff
@@ -13,27 +13,31 @@
...
     rank = MPI.COMM_WORLD.Get_rank()
     U.make_session(num_cpu=1).__enter__()
     workerseed = seed + 10000 * rank
     set_global_seeds(workerseed)
     env = SoccerEnv(rank)
 
     def policy_fn(name, ob_space, ac_space):
-        return mlp_policy.MlpPolicy(name=name, ob_space=ob_space, ac_space=ac_space,
-                                    hid_size=64, num_hid_layers=2)
+        return kick_policy.KickPolicy(name=name, ob_space=ob_space, ac_space=ac_space,
```
Comment (luckeciano) [inline baselines/ppo1/run_soccer.py:26]: I changed this line to my case but maybe could be better maintain the MlpPolicy as default (It is easier for who is starting to use this code).

File: baselines/ppo1/run_soccer.py [hunk 8]
```diff
@@ -48,17 +52,33 @@ def main():
     boolean_flag(parser, 'save-model', default=True)
     boolean_flag(parser, 'load-model', default=False)
     parser.add_argument('--model-dir')
+    parser.add_argument('--timesteps_per_actorbatch', type=int, default=4096)
```
Comment (luckeciano) [inline baselines/ppo1/run_soccer.py:55]: As I said to you, I added PPO hypers as arguments to automatize random search - which is the best thing I did to improve reward in my experiments.

**Review by luckeciano — APPROVED** [review]
Well, I think everything important is here and I commented. Maybe we should plan refactor some parts to make more legible. 

I also want to add the other branches in master code, as soon as I can.

File: baselines/ppo1/mlp_policy.py (hunk 2 above)
Comment (alexandremuzio) [reply baselines/ppo1/mlp_policy.py:44]: Yeah! Thanks for that

File: baselines/ppo1/kick_policy.py (hunk 3 above)
Comment (alexandremuzio) [reply baselines/ppo1/kick_policy.py:12]: No worries... It is useful, we'll fix it as we go. Thanks!

File: baselines/ppo1/pposgd_simple.py (hunk 4 above)
Comment (alexandremuzio) [reply baselines/ppo1/pposgd_simple.py:13]: which paper was it? Do you remember? I think I've saw something related in a deepmind paper

File: baselines/ppo1/pposgd_simple.py (hunk 5 above)
Comment (alexandremuzio) [reply baselines/ppo1/pposgd_simple.py:146]: Cool! Thanks for the explanation.

File: baselines/ppo1/pposgd_simple.py (hunk 6 above)
Comment (alexandremuzio) [reply baselines/ppo1/pposgd_simple.py:155]: Nice idea!

File: baselines/ppo1/run_soccer.py (hunk 7 above)
Comment (alexandremuzio) [reply baselines/ppo1/run_soccer.py:26]: Makes sense
